primitives/x25519/x25519.go

```diff
@@ -44,12 +46,84 @@ import (
 )
 
 const (
-	// ScalarSize is the size of the scalar input to X25519.
+	// ScalarSize is the size, in bytes, of the scalar input to X25519.
 	ScalarSize = 32
-	// PointSize is the size of the point input to X25519.
+
+	// PointSize is the size, in bytes, of the point input to X25519.
 	PointSize = 32
+
+	// PrivateKeySize is the size, in bytes, of private keys as used in this package.
+	PrivateKeySize = ScalarSize
+
+	// PublicKeySize is the size, in bytes, of public keys as used in this package.
+	PublicKeySize = PointSize
+
+	// SharedSecretSize is the size, in bytes, of shared secrets generated by this package.
+	SharedSecretSize = PointSize
+
+	// seedSize is the size, in bytes, of the seeds as used in this package.
+	seedSize = 32
 )
 
+// PrivateKey is the type of X25519 private keys.
+type PrivateKey [PrivateKeySize]byte
+
+// Public returns the public key corresponding to the private key.
+func (priv *PrivateKey) Public() *PublicKey {
+	var pub PublicKey
+	ScalarBaseMult((*[PointSize]byte)(&pub), (*[ScalarSize]byte)(priv))
+	return &pub
+}
+
+// DiffieHellman performs a Diffie-Hellman key exchange between the private key and the given
+// public key to produce a shared secret.
+func (priv *PrivateKey) DiffieHellman(pub *PublicKey) *SharedSecret {
+	var sec SharedSecret
+	ScalarMult((*[PointSize]byte)(&sec), (*[ScalarSize]byte)(priv), (*[PointSize]byte)(pub))
+	return &sec
+}
+
+// PublicKey is the type of X25519 public keys.
+type PublicKey [PublicKeySize]byte
+
+// SharedSecret is the type of the result of a Diffie-Hellman key exchange.
+type SharedSecret [SharedSecretSize]byte
+
+func (ss *SharedSecret) IsZero() bool {
+	var zero SharedSecret
+	return subtle.ConstantTimeCompare(ss[:], zero[:]) == 1
+}
+
+// GeneratePrivateKey generates a private key using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GeneratePrivateKey(rand io.Reader) (*PrivateKey, error) {
+	if rand == nil {
+		rand = cryptorand.Reader
+	}
+
+	var seed [seedSize]byte
+	if _, err := io.ReadFull(rand, seed[:]); err != nil {
+		return nil, err
+	}
+
+	digest := sha512.Sum512_256(seed[:]) // Mitigate poor quality entropy.
+	bzero(seed[:])
```

Comment:
This is largely pointless with Go (and most other languages), omit to be in line with the rest of the package please (Look at what `Sum512_256` does to see that entropy being left in places is inevitable).